casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala

```diff
@@ -1,127 +1,184 @@
+      sigsF <- Ref.of[F, Set[Signature]](Set.empty)
     } yield
-      new ApproveBlockProtocol(block, requiredSigs, now, duration, interval, approvedBlockF, sigsF)
-
-  def blockApprovalPacketHandler[F[_]: Monad: ApproveBlockProtocol](
-      peer: PeerNode): PartialFunction[Packet, F[Option[Packet]]] =
-    Function
-      .unlift(packetToBlockApproval)
-      .andThen {
-        case ba: BlockApproval =>
-          for {
-            _ <- ApproveBlockProtocol[F].addApproval(ba)
-          } yield none[Packet]
-      }
-
-  private def packetToBlockApproval(msg: Packet): Option[BlockApproval] =
-    if (msg.typeId == transport.BlockApproval.id)
-      Try(BlockApproval.parseFrom(msg.content.toByteArray)).toOption
-    else None
+      new ApproveBlockProtocolImpl[F](block,
+                                      requiredSigs,
+                                      trustedValidators,
+                                      now,
+                                      duration,
+                                      interval,
+                                      sigsF)
+
+  private class ApproveBlockProtocolImpl[
+      F[_]: Sync: Capture: ConnectionsCell: TransportLayer: Log: Time: Timer: RPConfAsk: LastApprovedBlock](
+      val block: BlockMessage,
+      val requiredSigs: Int,
+      val trustedValidators: Set[ByteString],
+      val start: Long,
+      val duration: FiniteDuration,
+      val interval: FiniteDuration,
+      private val sigsF: Ref[F, Set[Signature]])
+      extends ApproveBlockProtocol[F] {
+    private implicit val logSource: LogSource = LogSource(this.getClass)
+
+    private val candidate                 = ApprovedBlockCandidate(Some(block), requiredSigs)
+    private val u                         = UnapprovedBlock(Some(candidate), start, duration.toMillis)
+    private val serializedUnapprovedBlock = u.toByteString
+    private val candidateHash             = PrettyPrinter.buildString(block.blockHash)
+    private val sigData                   = Blake2b256.hash(candidate.toByteArray)
+
+    def addApproval(a: BlockApproval): F[Unit] = {
+      val validSig = for {
+        c   <- a.candidate if c == this.candidate
+        sig <- a.sig if Validate.signature(sigData, sig)
+      } yield sig
+
+      val trustedValidator =
+        if (signedByTrustedValidator(a)) {
+          true.pure[F]
+        } else {
+          Log[F].warn(s"APPROVAL: Received BlockApproval from untrusted validator.") *> false
+            .pure[F]
+        }
+
+      val isValid = for {
+        validValidators <- trustedValidator
+      } yield validValidators && validSig.isDefined
+
+      val sender =
+        a.sig.fold("<Empty Signature>")(sig => Base16.encode(sig.publicKey.toByteArray))
+
+      FlatMap[F].ifM(isValid)(
+        sigsF.update(_ + validSig.get) *> Log[F].info(
+          s"APPROVAL: received block approval from $sender"),
+        Log[F].warn(s"APPROVAL: ignoring invalid block approval from $sender")
+      )
+    }
```

Comment:
Are you not a fan of giant for/yields?

```scala
    for {
      validValidators <- if (signedByTrustedValidator(a)) {
        true.pure[F]
      } else {
        Log[F].warn(s"APPROVAL: Received BlockApproval from untrusted validator.") *> false.pure[F]
      }
      validSig = for {
        c   <- a.candidate if c == this.candidate
        sig <- a.sig if Validate.signature(sigData, sig)
      } yield sig
      isValid = validValidators && validSig.isDefined
      sender = a.sig.fold("<Empty Signature>")(sig => Base16.encode(sig.publicKey.toByteArray))
      _ <- if (isValid) {
                sigsF.update(_ + validSig.get) *> Log[F].info(s"APPROVAL: received block approval from $sender")
              } else {
                Log[F].warn(s"APPROVAL: ignoring invalid block approval from $sender")
              }
    } yield ()
```

Note: not tested

Reply:
Not really, but I can change if you want. I don't know why it says "not tested" because there is a test for that
```
  it should "not accept BlockApproved messages signed by not trusted validators" in {
```

Reply:
I meant that the code snippet I posted for you isn't tested. But yeah ok - personally I find it easier to read but if you disagree then maybe it isn't true in for a lot of others.